Ask for type of experiment

jedheaj314 opened the pull request:
Fixes issue #7 

Changes:
1. Ask for type of experiment
2. Refactor so that new type of experiments can be added in future
3. Added unit tests (Previous unit tests are as they were just in separate files)

jedheaj314 committed:
Refactor and add type choices

jedheaj314 committed:
Rename new_experiment and its test

jedheaj314 committed:
Merge branch 'main' into jedheaj314/features/issue-7/ask-tool-selection

jedheaj314 committed:
Add unit tests and delete redendunt files

jedheaj314 committed:
Add some emojis

jedheaj314 committed:
Remove env var print

jedheaj314 commented:
@macromania Ready for review :)

jedheaj314 committed:
Add time out to communicate function

macromania commented on src/experiment.py:
```diff
@@ -0,0 +1,22 @@
+        self.name = name
+
+    @abstractmethod
+    def create_resources(self):
```
From an abstraction point of view, maybe our interface can be just `Experiment.create`?

Creating the resources needed, which command to run etc seem to be implementation details that may not need to be exposed?

macromania commented on src/experiment_manager.py:
```diff
@@ -0,0 +1,119 @@
+from src.prompt_flow_experiment import PromptFlowExperiment
+
+
+class ExperimentFactory:
```
This reads too much Java to me 🤣

I can also suggest implementing a strategy pattern instead which makes it simpler and a bit easier to scale going forward.

It can be achieved inside the `ExperimentManager`. Although `ExperimentManager` is also very java-ish. Naming is hard.. 

Some examples of naming conventions from the Python standard library that could inspire a more pythonic name for `ExperimentManager`:

**ConfigParser**: A class for handling configuration files.
**FileHandler**: A class for handling file operations in the logging module.
**ThreadPoolExecutor**: A class for managing a pool of threads in the concurrent.futures module.
**BaseHTTPRequestHandler**: A class for handling HTTP requests in the http.server module.

I may go with `ExperimentHandler` in this sense..

Let's go back to implementing strategy pattern in `ExperimentHandler`. We can get rid of `ExperimentFactory` all together and go with a single class. See below `ExperimentHandler`.

```python
class Experiment(ABC):
    def __init__(self, name):
        self.name = name

    @abstractmethod
    def create(self):
        pass

class PromptFlowExperiment(Experiment):
    def create(self):
        print(f"Running PromptFlow experiment: {self.name}")

# Placeholder classes for other experiment types
class JupyterNotebookExperiment(Experiment):
    def create(self):
        raise NotImplementedError("Jupyter Notebook experiment not implemented yet.")

class PromptyExperiment(Experiment):
    def create(self):
        raise NotImplementedError("Prompty experiment not implemented yet.")

class PythonExperiment(Experiment):
    def create(self):
        raise NotImplementedError("Python experiment not implemented yet.")


class ExperimentHandler:
    _experiments = {
        ExperimentType.PROMPT_FLOW: PromptFlowExperiment,
        ExperimentType.JUPYTER_NOTEBOOK: JupyterNotebookExperiment,
        ExperimentType.PROMPTY: PromptyExperiment,
        ExperimentType.PYTHON: PythonExperiment,
    }
 
   def __init__(self):
   ####

   def _check_and_connect_virtual_env(self)
   #####

    @classmethod
    def create(cls, experiment_type, name):

        try:
            print("🔥 Welcome to the Prompt Ignite!")

            name = self._get_experiment_name()
            experiment_type = self._get_experiment_type()

           if experiment_type not in cls._experiments:
                raise ValueError(f"Unsupported experiment type: {experiment_type}")

            experiment = cls._experiment_classes[experiment_type](name)
            experiment.create()

            print("🔥 Experiment setup complete! 🚀")

        except NotImplementedError:
            print("🛠️ This setup is not implemented yet!")
            return
        except Exception as e:
            print(f"❌ Oops! Something went went wrong. {e}")
            return

````

jedheaj314 replied:
This makes sense. Thanks. I have changed the code to reflect above. :)

macromania reviewed (changes requested):
Naming and pattern suggestions...

macromania commented:
Also, please rebase your branch onto main so we can get the CI action running for this PR too :)

jedheaj314 committed:
Expose create function only

jedheaj314 committed:
Merge branch 'main' into jedheaj314/features/issue-7/ask-tool-selection

jedheaj314 committed:
Remove not required unit tests files

jedheaj314 committed:
Add enw line

jedheaj314 commented:
@macromania Thanks for the review. I have updated the code (rebased and updated naming conventions and other changes)
Could you please rereview when you get time :)

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+from enum import Enum
+
+
+class ExperimentType(Enum):
```
Let's move this into `src/entities.py`

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+    PYTHON = 'Hello Python'
+
+
+class Experiment(ABC):
```
Let's move this into `src/entities.py`

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+        pass
+
+
+class PromptFlowExperiment(Experiment):
```
Let's move this into `src/experiments/prompt_flow.py`

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+        return process.returncode
+
+
+class JupyterNotebookExperiment(Experiment):
```
Let's move this into `src/experiments/jupyter_notebook.py`

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+        raise NotImplementedError("Jupyter Notebook experiment not implemented yet.")
+
+
+class PromptyExperiment(Experiment):
```
Let's move this into `src/experiments/prompty.py`

macromania commented on src/experiment_handler.py:
```diff
@@ -0,0 +1,185 @@
+        raise NotImplementedError("Prompty experiment not implemented yet.")
+
+
+class PythonExperiment(Experiment):
```
Let's move this into `src/experiments/pure_python.py`

macromania reviewed (commented):
small code organization suggestions.

jedheaj314 committed:
Fix env variable test error

jedheaj314 committed:
Restructure

jedheaj314 committed:
Remove unused references

jedheaj314 committed:
Add entitties.py

macromania reviewed (approved):
Excellent work, thanks 🙏🏼
